src/IrcConnection/IrcConnection.cs

```diff
@@ -1050,17 +1050,65 @@ private void _OnConnectionError(object sender, EventArgs e)
             } catch (ConnectionException) {
             }
         }
-        
+
+        /// <summary>
+        /// 
+        /// </summary>
+        private abstract class FiniteThread
```

Comment:
I like your approach to generalize the code between read and write thread, but I am a bit worried it introduces regressions related to this comment:

                    // make sure we close the stream after the thread is gone, else
                    // the thread will think the connection is broken!
    #if LOG4NET
                    _Logger.Debug("Stop(): joining thread...");
    #endif
                    _Thread.Join();

See https://github.com/meebey/SmartIrc4net/blob/master/src/IrcConnection/IrcConnection.cs#L1108